File: src/main/java/net/hellomouse/progression_change/registries/ProgressionModRecipeRegistry.java
package net.hellomouse.progression_change.registries;

import net.hellomouse.progression_change.ProgressionMod;
import net.hellomouse.progression_change.recipe.StoneToCobbleRecipe;
import net.minecraft.recipe.RecipeSerializer;
import net.minecraft.recipe.RecipeType;
import net.minecraft.registry.RegistryKeys;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.RegistryObject;

public class ProgressionModRecipeRegistry {
    public static final DeferredRegister<RecipeType<?>> TYPE_DEF_REG = DeferredRegister.create(RegistryKeys.RECIPE_TYPE, ProgressionMod.MODID);
    public static final DeferredRegister<RecipeSerializer<?>> DEF_REG = DeferredRegister.create(RegistryKeys.RECIPE_SERIALIZER, ProgressionMod.MODID);
    public static final RegistryObject<RecipeType<StoneToCobbleRecipe>> BLOCK_TO_BLOCK_TYPE = TYPE_DEF_REG.register("block_to_block", () -> new RecipeType<>() {
    });
    public static final RegistryObject<RecipeSerializer<StoneToCobbleRecipe>> BLOCK_TO_BLOCK = DEF_REG.register("block_to_block", StoneToCobbleRecipe.Serializer::new);
}

File: src/main/kotlin/net/hellomouse/xeno_early_start/registries/ProgressionModRecipeRegistry.kt
package net.hellomouse.xeno_early_start.registries

import net.hellomouse.xeno_early_start.ProgressionMod
import net.hellomouse.xeno_early_start.recipe.StoneToCobbleRecipe
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.recipe.RecipeType
import net.minecraft.registry.RegistryKeys
import net.minecraftforge.registries.DeferredRegister
import net.minecraftforge.registries.RegistryObject
import java.util.function.Supplier

object ProgressionModRecipeRegistry {
    val TYPE_DEF_REG: DeferredRegister<RecipeType<*>> =
        DeferredRegister.create(RegistryKeys.RECIPE_TYPE, ProgressionMod.Companion.MODID)
    val DEF_REG: DeferredRegister<RecipeSerializer<*>> =
        DeferredRegister.create(RegistryKeys.RECIPE_SERIALIZER, ProgressionMod.Companion.MODID)

    @JvmField
    val BLOCK_TO_BLOCK_TYPE: RegistryObject<RecipeType<StoneToCobbleRecipe>> =
        TYPE_DEF_REG.register("block_to_block", Supplier {
            object : RecipeType<StoneToCobbleRecipe> {
            }
        })

    @JvmField
    val BLOCK_TO_BLOCK: RegistryObject<RecipeSerializer<StoneToCobbleRecipe>> =
        DEF_REG.register(
            "block_to_block",
            Supplier { StoneToCobbleRecipe.Serializer() })
}
